README.md
#Hot Threads Plugin for Vanilla 2.0/2.1b

##Description
Hot Threads displays a list of the Hot Discussions, base on the amount of Views and Comments they received.

The list is displayed in a widget, inside the Panel, and a "Hot" lnk is added to Vanilla Discussions menu.

##Usage
The plugin works out of the box with its default configuration. Changing the settings is, however, very easy.

The following settings are available to configure:

###General
Maximum amount of entries to display. The Hot Threads widget will display this amount of entries.
Page display filter. This filter allows to display the Hot Threads widget only when the User is visiting a certain type of pages.
Auto-Refresh Delay. The Hot Threads Widget can refresh its content automatically. This setting indicates how many seconds the Widget should wait between each refresh. Setting it to zero disables the Auto-Refresh.

###Thresholds
* **Age Threshold**. This value is used to make sure that old Hot Discussions do not keep appearing before newer ones. Whenever a Discussion is older than the specified amount of days, it will get a lower priority in the list.
* **Comment Count Threshold**. This value indicates how many comments a Discussion must have received to be considered "hot". Set it to zero to indicate "any amount".
* **View Count Threshold**. This value indicates how many views a Discussion must have to be considered "hot". Set it to zero to indicate "any amount".

You can find more details about how the Thresholds are processed in Plugin's Settings page.

##Requirements

* PHP 5.3+
* Vanilla 2.0/2.1b (see Notes)

##Notes
Plugin has been tested on Vanilla 2.1b and it should be compatibile with it. However, since Vanilla 2.1 is still in Beta, compatibility cannot be guaranteed. For the same reason, we provide limited assistance for installations in such environment.

##Credits
Thanks to [Zach](http://vanillaforums.org/profile/38268/hgtonight), from Vanilla Community, for the ideas behind this plugin. You can see his work on [his website](http://www.daklutz.com/).


HotThreads/CHANGELOG.md
#Hot Threads Plugin for Vanilla Forums 2.0/2.1 - Change log

####14.09.04
* Fixed bug in parsing of the arguments passed with a `/hotthreads` request.

####14.08.19
* Fixed calculation of the total "hot discussions" used for paging purposes.

####14.01.24
* Changed licence to GPLv3.

####13.04.02
* Corrected issue with permissions. Previous versions allowed only SuperAdmin to access Plugin's settings.

####13.03.07
* Added option to completely hide the Hot Threads Widget if there are no Discussions to show.

####13.02.20
* Added Age Threshold to prioritise newer Hot Discussions.

####13.02.15
* Added readme.html.

####13.02.07
* Initial release.


HotThreads/js/hotthreadswidget.js
$(document).ready(function(){
	/**
	 * Updates the content of the Hot Threads Widget.
	 */
	function UpdateHotThreadsWidget() {
		var HotThreadsListURL = gdn.url('/plugin/hotthreads/getwidgetcontent');

		$.ajax({
			url: HotThreadsListURL,
			global: false,
			dataType: 'html',
			success: function(Data){
				var CurrentItems = $("#HotThreadsList_Items").html();
				// Update widget only if received data is different from the old one
				if(CurrentItems != Data) {
					// Hide current Widget content with a fade effect
					$("#HotThreadsList_Items").fadeOut('slow', function() {
						// Replace widget content and display it again, using the same fade effect
						$(this).html(Data);
						$("#HotThreadsList_Items").fadeIn('slow');
					});
				}

				// Set the auto refresh to occur again later
				setTimeout(UpdateHotThreadsWidget, gdn.definition('HotThreadsWidget_AutoUpdateDelay') * 1000);
			}
		});
	}

	// Hot Threads widget might not exist if there are no Hot Threads to display
	// and Admins chose to hide it when empty. In such case, there is no point in
	// trying to automatically update it
	var HotThreadsWidget = $('#HotThreadsList_Items');
	if(HotThreadsWidget) {
		// If Auto Update delay is greather than zero, it means it's enabled. In such
		// case, it starts the auto-refresh loop
		var AutoUpdateDelay = gdn.definition('HotThreadsWidget_AutoUpdateDelay');
		if(gdn.definition('HotThreadsWidget_AutoUpdateDelay') > 0) {
			setTimeout(UpdateHotThreadsWidget, gdn.definition('HotThreadsWidget_AutoUpdateDelay') * 1000);
		}
	}
});


HotThreads/lib/hotthreads.defines.php
<?php if (!defined('APPLICATION')) exit();

/**
 * Constants used by HotThreads Plugin.
 */

define('HOTTHREADS_PAGESET_ALL', 'all');
define('HOTTHREADS_PAGESET_DISCUSSIONS', 'discussions');
define('HOTTHREADS_PAGESET_ANNOUNCEMENTS', 'announcements');

// Default Configuration Settings

// Default number of hot threads to display
define('HOTTHREADS_DEFAULT_MAXENTRIES', 10);
// Default Views Threshold. A Discussion will be considered "hot" if it received
// at least this amount of views
define('HOTTHREADS_DEFAULT_VIEWSTHRESHOLD', 100);
// Default Comments Threshold. A Discussion will be considered "hot" if it
// received at least this amount of comments
define('HOTTHREADS_DEFAULT_COMMENTSTHRESHOLD', 10);
// Default Age Threshold to apply to Discussions to determine their order
// when they are displayed in the Hot Threads page and widget
define('HOTTHREADS_DEFAULT_AGETHRESHOLD', 30);
// Default delay for the automatic updating of the Hot Threads list, in seconds
define('HOTTHREADS_DEFAULT_AUTOUPDATEDELAY', 120);
// By default, hide Hot Threads widget if no Hot Threads are available
define('HOTTHREADS_DEFAULT_HIDEIFEMPTY', 1);

// Paths
define('HOTTHREADS_PLUGIN_PATH', PATH_PLUGINS . '/HotThreads');
define('HOTTHREADS_PLUGIN_LIB_PATH', HOTTHREADS_PLUGIN_PATH . '/lib');
define('HOTTHREADS_PLUGIN_CLASSES_PATH', HOTTHREADS_PLUGIN_LIB_PATH . '/classes');
define('HOTTHREADS_PLUGIN_MODULES_PATH', HOTTHREADS_PLUGIN_CLASSES_PATH . '/modules');

define('HOTTHREADS_PLUGIN_VIEWS_PATH', HOTTHREADS_PLUGIN_PATH . '/views');

// URLs
define('HOTTHREADS_PLUGIN_BASE_URL', '/plugin/hotthreads');
define('HOTTHREADS_GENERALSETTINGS_URL', HOTTHREADS_PLUGIN_BASE_URL . '/settings');
define('HOTTHREADS_PAGE_URL', '/discussions/hotthreads');

// Return Codes
define('HOTTHREADS_OK', 0);


HotThreads/views/hotthreads_generalsettings_view.php
<?php if (!defined('APPLICATION')) exit();
?>
<div class="HotThreadsPlugin">
	<div class="Header">
		<h1><?php echo T($this->Data['Title']); ?></h1>
		<div class="Info">
			<?php echo T($this->Data['PluginDescription']); ?>
		</div>
	</div>
	<div class="Content">
		<?php
			echo $this->Form->Open();
			echo $this->Form->Errors();
		?>
		<fieldset>
			<legend>
				<h3><?php echo T('General Settings'); ?></h3>
			</legend>
			<ul>
				<li><?php
					echo $this->Form->Label(T('Maximum amount of entries to display'),
																		'Plugin.HotThreads.Count');
					echo Wrap(T('The Hot Threads widget will display this amount of entries.'),
										'div',
										array('class' => 'Info',));
					echo $this->Form->Textbox('Plugin.HotThreads.MaxEntriesToDisplay');
				?></li>
				<li><?php
					echo $this->Form->Label(T('On which pages would you like to display the widget?'),
																		'Plugin.HotThreads.DisplayPageSet');
					echo $this->Form->DropDown('Plugin.HotThreads.DisplayPageSet',array(
								HOTTHREADS_PAGESET_ALL => T('Discussions and Announcements'),
								HOTTHREADS_PAGESET_DISCUSSIONS => T('Only Discussions'),
								HOTTHREADS_PAGESET_ANNOUNCEMENTS => T('Only Announcements'),
					));
				?></li>
				<li><?php
					echo $this->Form->Label(T('Refresh Hot Threads widget every X seconds'),
																	'Plugin.HotThreads.AutoUpdateDelay');
					echo Wrap(T('Set this value to zero to disable automatic refresh.'),
										'div',
										array('class' => 'Info',));
					echo $this->Form->Textbox('Plugin.HotThreads.AutoUpdateDelay');
				?></li>
				<li><?php
					echo $this->Form->Checkbox('Plugin.HotThreads.HideIfEmpty', T('Hide sidebar Widget completely if no Hot Threads are found'));
				?></li>
			</ul>
		</fieldset>
		<fieldset>
			<legend>
				<h3><?php echo T('Thresholds'); ?></h3>
				<?php
					echo Wrap(T('Thresholds are used to determine if a thread (discussion) ' .
											'should be considered "hot". ' .
											'Please note that a thread only needs to be over one ' .
											'threshold to be considered "hot". That is, a thread must ' .
											'have <strong>either</strong> a certain amount of Views, ' .
											'<strong>or</strong> a certain amount of Comments, ' .
											'<strong>or</strong> both.'),
										'div',
										array('class' => 'Info',));
				?>
			</legend>
			<div class="Thresholds Help Aside"><?php
				echo Wrap(T('How do Thresholds work?'),
									'h4',
									array('class' => 'Title',));
				echo Wrap(T('Thresholds are used to determine if a thread (discussion) ' .
										'should be considered "hot", and also how hot they are.'),
									'div',
									array('class' => 'Info',));
				echo Wrap(T('The following example assumes a <strong>View Threshold</strong> of 100 and ' .
										'a <strong>Comments Threshold</strong> of 10.'),
									'div',
									array('class' => 'Info',));
				echo "<ul>\n";
				echo Wrap(T('Discussion X and Y have both 10 Views and 1 Comment. They are <strong>not</strong> hot.'),
									'li');
				echo Wrap(T('Discussion X reaches 100 Views, but still 1 Comment. Discussion X ' .
										'<strong>is hot</strong>, since it passed the Views Threshold.'),
									'li');
				echo Wrap(T('Discussion Y reaches 20 Views and 10 Comments. Discussion Y ' .
										'<strong>is hot</strong>, since it passed the Comments Threshold.'),
									'li');
				echo Wrap(T('The Hot Threads Widget and Page will now display Discussions X and Y. ' .
										'Discussion Y will appear <strong>before</strong> Discussion X, because ' .
										'Comments have more weight in determining the "hotness" of a Discussion.'),
									'li');
				echo "</ul>\n";
			?></div>
			<ul>
				<li><?php
					echo $this->Form->Label(T('Age Threshold (days)'),
																	'Plugin.HotThreads.AgeThreshold');
					echo Wrap(T('This value is used to make sure that old Hot Discussions do not keep ' .
											'appearing before newer ones. Whenever a Discussion is older than the ' .
											'specified amount of days, it will get a lower priority in the list.'),
										'div',
										array('class' => 'Info',));
					echo $this->Form->Textbox('Plugin.HotThreads.AgeThreshold');
				?></li>
				<li><?php
					echo $this->Form->Label(T('Comment Count Threshold'),
																	'Plugin.HotThreads.CommentsThreshold');
					echo Wrap(T('This value indicates how many comments a Discussion must have ' .
											'received to be considered "hot".'),
										'div',
										array('class' => 'Info',));
					echo $this->Form->Textbox('Plugin.HotThreads.CommentsThreshold');
				?></li>
				<li><?php
					echo $this->Form->Label(T('View Count Threshold'),
																	'Plugin.HotThreads.ViewsThreshold');
					echo Wrap(T('This value indicates how many views a Discussion must have ' .
											'to be considered "hot".'),
										'div',
										array('class' => 'Info',));
					echo $this->Form->Textbox('Plugin.HotThreads.ViewsThreshold');
				?></li>
			</ul>
		</fieldset>
		<?php
			echo $this->Form->Close('Save');
		?>
	</div>
	<div class="Credits">
		<h3>Credits</h3>
		<p>
			Thanks to <a href="http://vanillaforums.org/profile/38268/hgtonight" title="Zach's profile on Vanilla Community">Zach</a>, from Vanilla Community, for the ideas behind this plugin. You can see his work on <a href="http://www.daklutz.com" title="Zach's Website">his website</a>.
		</p>
	</div>
</div>


HotThreads/class.hotthreads.plugin.php
<?php if(!defined('APPLICATION')) exit();

/* Copyright 2013 Diego Zanella (support@pathtoenlightenment.net)
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 3, as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   GPL3: http://www.gnu.org/licenses/gpl-3.0.txt
*/

// File hotthreads.defines.php must be included by manually specifying the whole
// path. It will then define some shortcuts for commonly used paths.
require(PATH_PLUGINS . '/HotThreads/lib/hotthreads.defines.php');

// Define the plugin:
$PluginInfo['HotThreads'] = array(
	'Name' => 'Hot Threads',
	'Description' => 'Displays a list of "hot" discussions, i.e. the ones with most views and/or comments.',
	'Version' => '14.09.04',
	'RequiredApplications' => array('Vanilla' => '2.0'),
	'RequiredTheme' => FALSE,
	'HasLocale' => FALSE,
	'MobileFriendly' => TRUE,
	'SettingsUrl' => '/plugin/hotthreads',
	'SettingsPermission' => 'Garden.Settings.Manage',
	'Author' => 'D.Zanella',
	'AuthorEmail' => 'diego@pathtoenlightenment.net',
	'AuthorUrl' => 'http://dev.pathtoenlightenment.net',
	'RegisterPermissions' => array('Plugins.HotThreads.Manage',),
);

/**
 * Displays a Widget and a Page with a list of Hot Threads, based on their View
 * Count and their Comment Count.
 */
class HotThreadsPlugin extends Gdn_Plugin {
	/**
	 * Class constructor.
	 *
	 * @return HotThreadsPlugin.
	 */
	public function __construct() {
		parent::__construct();
	}

	/**
	 * Given the name of a Page Set, it returns all the pages included into it.
	 * It's mainly used internally, to determine if the Hot Threads Widget should
	 * be displayed, based on the Controller that handles the requests.
	 *
	 * @param string PageSetName The name of the Page Set to retrieve.
	 * @return array An array containing a list of the pages included in the Page
	 * Set.
	 */
	private function GetDisplayPages($PageSetName = HOTTHREADS_PAGESET_ALL) {
		$PageSets = array(
			HOTTHREADS_PAGESET_DISCUSSIONS => array(
				// Note that the two lines below are not duplicate. One is "discussion",
				// the other is discussionS (plural)
				'discussioncontroller',
				'discussionscontroller',
				'categoriescontroller',
			),
			HOTTHREADS_PAGESET_ANNOUNCEMENTS => array(
				'profilecontroller',
				'activitycontroller'
			)
		);

		// Produce an "all" entry, containing all the pages of the other pagesets
		$PageSets[HOTTHREADS_PAGESET_ALL] = array_merge($PageSets[HOTTHREADS_PAGESET_DISCUSSIONS],
																										$PageSets[HOTTHREADS_PAGESET_ANNOUNCEMENTS]);

		return GetValue($PageSetName, $PageSets);
	}


	/**
	 * Determines if the Hot Threads widget should be displayed, by checking if
	 * the requesting Controller matches the pages where the widget was enabled.
	 *
 	 * @param Controller Sender Sending controller instance.
 	 * @return bool True, if the Widget should be displayed, False otherwise.
 	 */
	private function ShouldDisplayWidget($Sender) {
		// Get the config and controller name for comparison
		$DisplayPageSet = C('Plugin.HotThreads.DisplayPageSet', HOTTHREADS_PAGESET_ALL);

		return InArrayI($Sender->ControllerName, $this->GetDisplayPages($DisplayPageSet));
	}

	// Create a method called "HotThreadsPlugin" on the PluginController
   public function PluginController_HotThreads_Create($Sender) {
		$Sender->Title('Hot Threads Plugin');
		$Sender->AddSideMenu('plugin/hotthreadslist');

		// Prepare a form to be used by call handlers
		$Sender->Form = new Gdn_Form();
		$this->Dispatch($Sender, $Sender->RequestArgs);
	}

	/**
	 * Renders the Hot Threads page.
	 * This method is an almost exact copy of DiscussionControllers::Index(), with
	 * the exception that it doesn't load Announcements. Also, it loads only
	 * Discussions with a certain amount of Views or Comments.
	 *
 	 * @param Controller Sender Sending controller instance.
	 * @param int Page The page to display (used by Pager).
	 */
	private function ShowHotThreads($Sender, $Page = '0') {
		// Determine offset from $Page
		list($Page, $Limit) = OffsetLimit($Page, C('Vanilla.Discussions.PerPage', 30));
		$Sender->CanonicalUrl(Url(ConcatSep('/', 'discussions', PageNumber($Page, $Limit, TRUE, FALSE)), TRUE));

		// Validate $Page
		if (!is_numeric($Page) || $Page < 0) {
			$Page = 0;
		}

		// Setup head.
		if (!$Sender->Data('Title')) {
			$Sender->Title(T('Hot Discussions'));
		}

		// Add modules
		$Sender->AddModule('NewDiscussionModule');
		$Sender->AddModule('CategoriesModule');
		$Sender->AddModule('BookmarkedModule');

		// Set criteria & get discussions data
		$Sender->SetData('Category', FALSE, TRUE);
		$DiscussionModel = new DiscussionModel();
		$DiscussionModel->Watching = TRUE;


		/* Half a day is added to the Age Threshold because we want to give maximum
		 * priority to all the Discussion whose age is less than, or equal to the
		 * configured value. Due to the calculation performed to determine priority,
		 * a Discussion whose age is equal to the threshold would get a priority of
		 * "1", rather than the correct "0".
		 * By adding half a day, the calculation will produce a result which is less
		 * than 1, giving the Discussions the correct priority.
		 */
		$AgeThreshold = C('Plugin.HotThreads.AgeThreshold', HOTTHREADS_DEFAULT_AGETHRESHOLD);
		$AgeThreshold = (int)$AgeThreshold + 0.5;

		// Retrieve the total amount of "hot discussions", for paging purposes
		$DiscussionModel->SQL
			->Select('d.DiscussionID', 'COUNT(%s)', 'HotDiscussionsCount')
			->BeginWhereGroup()
			->Where('d.CountViews >=', C('Plugin.HotThreads.ViewsThreshold',
																	 HOTTHREADS_DEFAULT_VIEWSTHRESHOLD))
			->OrWhere('d.CountComments >=', C('Plugin.HotThreads.CommentsThreshold',
																				HOTTHREADS_DEFAULT_COMMENTSTHRESHOLD))
			->EndWhereGroup();

		// Get Discussion Count
		$DiscussionsCountData = $DiscussionModel->Get()->FirstRow();
		$CountDiscussions = ($DiscussionsCountData === false) ? 0 : $DiscussionsCountData->HotDiscussionsCount;
		$Sender->SetData('CountDiscussions', $CountDiscussions);

		// Filter the Discussions, keeping only the ones with a certain amount of
		// Views or Comments
		$DiscussionModel->SQL
			->Select('FLOOR((TO_DAYS(NOW())-TO_DAYS(d.DateLastComment))/' . $AgeThreshold . ')', '', 'AgeWeight')
			->Select('(TO_DAYS(NOW())-TO_DAYS(d.DateLastComment))', '', 'Age')
			->BeginWhereGroup()
			->Where('d.CountViews >=', C('Plugin.HotThreads.ViewsThreshold',
																	 HOTTHREADS_DEFAULT_VIEWSTHRESHOLD))
			->OrWhere('d.CountComments >=', C('Plugin.HotThreads.CommentsThreshold',
																				HOTTHREADS_DEFAULT_COMMENTSTHRESHOLD))
			->EndWhereGroup()
			->OrderBy('AgeWeight', 'asc')
			->OrderBy('d.CountComments', 'desc')
			->OrderBy('d.CountViews', 'desc');

		// Get Discussions
		$Sender->DiscussionData = $DiscussionModel->Get($Page, $Limit);

//		var_dump($Sender->DiscussionData);
//		die();

		$Sender->SetData('Discussions', $Sender->DiscussionData, TRUE);
		$Sender->SetJson('Loading', $Page . ' to ' . $Limit);

		// Build a pager
		$PagerFactory = new Gdn_PagerFactory();
		$Sender->EventArguments['PagerType'] = 'Pager';
		$Sender->FireEvent('BeforeBuildPager');
		$Sender->Pager = $PagerFactory->GetPager($Sender->EventArguments['PagerType'], $this);
		$Sender->Pager->ClientID = 'Pager';
		$Sender->Pager->Configure(
			$Page,
			$Limit,
			$CountDiscussions,
			'discussions/%1$s'
		);
		$Sender->SetData('_PagerUrl', 'discussions/hotthreads/{Page}');
		$Sender->SetData('_Page', $Page);
		$Sender->SetData('_Limit', $Limit);
		$Sender->FireEvent('AfterBuildPager');

		// Deliver JSON data if necessary
		if($Sender->DeliveryType() != DELIVERY_TYPE_ALL) {
			$Sender->SetJson('LessRow', $Sender->Pager->ToString('less'));
			$Sender->SetJson('MoreRow', $Sender->Pager->ToString('more'));
			$Sender->View = 'discussions';
		}

		// Set a definition of the user's current timezone from the db. jQuery
		// will pick this up, compare to the browser, and update the user's
		// timezone if necessary.
		$CurrentUser = Gdn::Session()->User;
		if (is_object($CurrentUser)) {
			$ClientHour = $CurrentUser->HourOffset + date('G', time());
			$Sender->AddDefinition('SetClientHour', $ClientHour);
		}

		// Render default view (discussions/index.php)
		$Sender->View = 'index';
		$Sender->Render();
	}

	/**
	 * Loads and configures the Hot Threads module, which will generate the HTML
	 * for the Hot Threads widget in the Sidebar.
	 *
 	 * @param Controller Sender Sending controller instance.
 	 * @return HotThreadsListModule An instance of the module.
 	 */
	private function LoadHotThreadsModule($Sender) {
		// Include Hot Threads List module file
		//include_once(HOTTHREADS_PLUGIN_MODULES_PATH . '/class.hotthreadslist.module.php');

		$HotThreadsPluginModule = new HotThreadsListModule($Sender);
		$HotThreadsPluginModule->LoadData(
			C('Plugin.HotThreads.MaxEntriesToDisplay'),
			C('Plugin.HotThreads.ViewsThreshold'),
			C('Plugin.HotThreads.CommentsThreshold'),
			C('Plugin.HotThreads.AgeThreshold')
		);
		return $HotThreadsPluginModule;
	}

	/**
	 * Handler of global "Render_Before" event.
	 * It initialises and renders the module displaying the hot threads.
	 *
   * @param Controller Sender Sending controller instance.
	 */
	public function Base_Render_Before($Sender) {
		$Sender->AddCssFile('hotthreads.css', 'plugins/HotThreads/design/css');

		if(!$this->ShouldDisplayWidget($Sender)) {
			return;
		}

		// TODO Read configuration to load multiple Discussion list widgets in a specific order

		// Load the module that will render the Hot Threads widget and add it to the
		// modules list
		$HotThreadsPluginModule = $this->LoadHotThreadsModule($Sender);
		$Sender->AddModule($HotThreadsPluginModule);

		// If Auto Update is enabled (delay greater than zero), load and configure
		// the related JavaScript file
		if(($AutoUpdateDelay = C('Plugin.HotThreads.AutoUpdateDelay', HOTTHREADS_DEFAULT_AUTOUPDATEDELAY)) > 0) {
			$this->LoadWidgetAutoRefreshScript($Sender, $AutoUpdateDelay);
		}
	}

	/**
	 * Loads and configures the JavaScript file containing the script that will
	 * automatically refresh the Hot Threads Widget.
	 *
   * @param Controller Sender Sending controller instance.
	 * @param int AutoUpdateDelay The delay between each refresh, in seconds.
	 */
	protected function LoadWidgetAutoRefreshScript($Sender, $AutoUpdateDelay = HOTTHREADS_DEFAULT_AUTOUPDATEDELAY) {
		// Load the JS file that will handle Widget's auto-refresh
		$Sender->AddJsFile('hotthreadswidget.js', 'plugins/HotThreads/js');

		// Expose the Auto Update interval to the front-end, so that it can be used
		// by the JS script
		$Sender->AddDefinition('HotThreadsWidget_AutoUpdateDelay', $AutoUpdateDelay);
	}

	/**
	 * Renders the HTML produced by the Hot Threads Module. This method is mainly
	 * used to updated the widget "on the fly" via Ajax calls.
	 */
	public function Controller_GetWidgetContent($Sender) {
		// TODO Parse Request to determine which data to load and display in the widget

		// Load the module that will render the Hot Threads widget and output the
		// HTML it generates
		$HotThreadsPluginModule = $this->LoadHotThreadsModule($Sender);
		echo $HotThreadsPluginModule->RenderListItems();
	}

	/**
	 * Add Controller to display Hot Threads.
	 *
	 * @param Gdn_Controller Sender Sending controller instance.
	 * @param array Args The arguments passed with the request.
	 */
	public function DiscussionsController_HotThreads_Create($Sender, $Args) {
		// Replace standard View with the "Hot Threads" view
		$this->ShowHotThreads($Sender, GetValue(0, $Args, 'p1'));
	}

	/**
	 * Vanilla 2.0 Event Handler.
	 * Adds a "Hot Threads" link to tabs in Index page.
	 *
	 * @param Controller Sender Sending controller instance.
	 */
	public function Base_AfterDiscussionTabs_Handler($Sender) {
		// TODO Review code to display a Sprite in Vanilla 2.1
		$CssClass = $Sender->RequestMethod == 'hotthreads' ? 'Active' : '';
		echo Wrap(Anchor(T('Hot'),
										 '/discussions/hotthreads',
										 'TabLink'),
							'li',
							array('class' => 'HotThreads ' . $CssClass)
						 );
	}

	/**
	 * Vanilla 2.1 Event Handler.
	 * Calls HotThreadsPlugin::Base_AfterDiscussionTabs_Handler().
	 *
	 * @see HotThreadsPlugin::Base_AfterDiscussionTabs_Handler().
	 */
	public function DiscussionsController_AfterDiscussionFilters_Handler($Sender) {
		return $this->Base_AfterDiscussionTabs_Handler($Sender);
	}

	/**
	 * Vanilla 2.1 Event Handler.
	 * Calls HotThreadsPlugin::Base_AfterDiscussionTabs_Handler().
	 *
	 * @see HotThreadsPlugin::Base_AfterDiscussionTabs_Handler().
	 */
	public function DiscussionController_AfterDiscussionFilters_Handler($Sender) {
		return $this->Base_AfterDiscussionTabs_Handler($Sender);
	}

	/**
	 * Renders the default page.
	 *
   * @param Controller Sender Sending controller instance.
	 */
	public function Controller_Index($Sender) {
		$this->Controller_Settings($Sender);
	}

	/**
	 * Set validation rules for Plugin's configuration settings.
	 *
	 * @param Gdn_Validation Validation The Validation object to which the rules
	 * will be added.
	 */
	private function _SetConfigValidationRules(Gdn_Validation $Validation) {
		$Validation->ApplyRule('Plugin.HotThreads.DisplayPageSet', 'Required');

		$Validation->ApplyRule('Plugin.HotThreads.AutoUpdateDelay', 'Required');
		$Validation->ApplyRule('Plugin.HotThreads.AutoUpdateDelay', 'Integer');

		$Validation->ApplyRule('Plugin.HotThreads.MaxEntriesToDisplay', 'Required');
		$Validation->ApplyRule('Plugin.HotThreads.MaxEntriesToDisplay', 'Integer');

		// Thresholds
		$Validation->ApplyRule('Plugin.HotThreads.AgeThreshold', 'Required');
		$Validation->ApplyRule('Plugin.HotThreads.AgeThreshold', 'Integer');
		$Validation->ApplyRule('Plugin.HotThreads.ViewsThreshold', 'Required');
		$Validation->ApplyRule('Plugin.HotThreads.ViewsThreshold', 'Integer');
		$Validation->ApplyRule('Plugin.HotThreads.CommentsThreshold', 'Required');
		$Validation->ApplyRule('Plugin.HotThreads.CommentsThreshold', 'Integer');
	}

	/**
	 * Renders the Settings page.
	 *
   * @param Controller Sender Sending controller instance.
	 */
	public function Controller_Settings($Sender) {
		$Sender->Permission('Vanilla.Settings.Manage');
		$Sender->SetData('PluginDescription',$this->GetPluginKey('Description'));

		$Validation = new Gdn_Validation();
		$ConfigurationModel = new Gdn_ConfigurationModel($Validation);
		$ConfigurationModel->SetField(array(
			'Plugin.HotThreads.HideIfEmpty' => HOTTHREADS_DEFAULT_HIDEIFEMPTY,
			'Plugin.HotThreads.DisplayPageSet' => HOTTHREADS_PAGESET_ALL,
			'Plugin.HotThreads.AutoUpdateDelay'	=> HOTTHREADS_DEFAULT_AUTOUPDATEDELAY,
			'Plugin.HotThreads.MaxEntriesToDisplay'	=> HOTTHREADS_DEFAULT_MAXENTRIES,
			'Plugin.HotThreads.AgeThreshold'	=> HOTTHREADS_DEFAULT_AGETHRESHOLD,
			'Plugin.HotThreads.ViewsThreshold'	=> HOTTHREADS_DEFAULT_VIEWSTHRESHOLD,
			'Plugin.HotThreads.CommentsThreshold'	=> HOTTHREADS_DEFAULT_COMMENTSTHRESHOLD
		));

		// Set the model on the form.
		$Sender->Form->SetModel($ConfigurationModel);

		// If seeing the form for the first time...
		if($Sender->Form->AuthenticatedPostBack() === FALSE) {
			// Apply the config settings to the form.
			$Sender->Form->SetData($ConfigurationModel->Data);
		}
		else {
			// Set validation rules for configuration parameters
			$this->_SetConfigValidationRules($ConfigurationModel->Validation);

			$Saved = $Sender->Form->Save();

			if($Saved) {
				$Sender->InformMessage('<span class="InformSprite Sliders"></span>'.T("Your changes have been saved."),'HasSprite');
			}
		}

		// Render the Settings view
		$Sender->Render($this->GetView('hotthreads_generalsettings_view.php'));
	}

	/**
	 * Add a link to Plugin's configuration page to the dashboard menu.
	 *
   * @param Controller Sender Sending controller instance.
	 */
	public function Base_GetAppSettingsMenuItems_Handler($Sender) {
		$Menu = $Sender->EventArguments['SideMenu'];
		$Menu->AddLink('Add-ons', T('Hot Threads'), 'plugin/hotthreads', 'Garden.Settings.Manage');
	}

	/**
	 * Performs the initial Setup when the plugin is enabled.
	 */
	public function Setup() {
		// Set Plugin's default settings
		SaveToConfig('Plugin.HotThreads.HideIfEmpty', HOTTHREADS_DEFAULT_HIDEIFEMPTY);
		SaveToConfig('Plugin.HotThreads.AutoUpdateDelay', HOTTHREADS_DEFAULT_AUTOUPDATEDELAY);
		SaveToConfig('Plugin.HotThreads.MaxEntriesToDisplay', HOTTHREADS_DEFAULT_MAXENTRIES);
		SaveToConfig('Plugin.HotThreads.DisplayPageSet', HOTTHREADS_PAGESET_ALL);
		SaveToConfig('Plugin.HotThreads.AgeThreshold', HOTTHREADS_DEFAULT_AGETHRESHOLD);
		SaveToConfig('Plugin.HotThreads.ViewsThreshold', HOTTHREADS_DEFAULT_VIEWSTHRESHOLD);
		SaveToConfig('Plugin.HotThreads.CommentsThreshold', HOTTHREADS_DEFAULT_COMMENTSTHRESHOLD);
	}

	/**
	 * Performs a cleanup when the plugin is removed.
	 */
	public function Cleanup() {
		// Remove Plugin's settings
		RemoveFromConfig('Plugin.HotThreads.HideIfEmpty');
		RemoveFromConfig('Plugin.HotThreads.AutoUpdateDelay');
		RemoveFromConfig('Plugin.HotThreads.MaxEntriesToDisplay');
		RemoveFromConfig('Plugin.HotThreads.DisplayPageSet');
		RemoveFromConfig('Plugin.HotThreads.AgeThreshold');
		RemoveFromConfig('Plugin.HotThreads.ViewsThreshold');
		RemoveFromConfig('Plugin.HotThreads.CommentsThreshold');
	}
}
